src/socket/icmp.rs

```diff
@@ -244,13 +248,23 @@ impl<'a, 'b> IcmpSocket<'a, 'b> {
 
     /// Filter determining which packets received by the interface are appended to
     /// the given sockets received buffer.
-    pub(crate) fn accepts(&self, ip_repr: &IpRepr, icmp_repr: &Icmpv4Repr,
+    pub(crate) fn accepts(&self, ip_repr: &IpRepr, icmp_repr: &IcmpRepr,
                           cksum: &ChecksumCapabilities) -> bool {
         match (&self.endpoint, icmp_repr) {
             // If we are bound to ICMP errors associated to a UDP port, only
             // accept Destination Unreachable messages with the data containing
             // a UDP packet send from the local port we are bound to.
-            (&Endpoint::Udp(endpoint), &Icmpv4Repr::DstUnreachable { data, .. })
+            #[cfg(feature = "proto-ipv4")]
+            (&Endpoint::Udp(endpoint), &IcmpRepr::Ipv4(Icmpv4Repr::DstUnreachable { data, .. }))
+                    if endpoint.addr.is_unspecified() || endpoint.addr == ip_repr.dst_addr() => {
+                let packet = UdpPacket::new(data);
+                match UdpRepr::parse(&packet, &ip_repr.src_addr(), &ip_repr.dst_addr(), cksum) {
+                    Ok(repr) => endpoint.port == repr.src_port,
+                    Err(_) => false,
+                }
+            }
+            #[cfg(feature = "proto-ipv6")]
```

Comment:
I feel like these two cases ought to be unified, but not sure if this can be done cleanly.

Reply:
I could add a function to factorize the code, but I'm not sure it improves readability

Reply:
I was thinking of something like
```rust
match x {
  #[cfg(foo)] Y | #[cfg(bar)] Z => t
}
```
but I don't know if it works.

Reply:
It doesn't, I tried